### PR TITLE
Enhance Swagger Generation

### PR DESCRIPTION
**Pull request by crobibero**
Add Redoc as a swagger parser
Move docs to /api-docs/swagger and /api-docs/redoc 
Enable xmldoc files to swagger gen
Enable json string enum converter

Part of #2872

### CODE REVIEW

**Comment by crobibero** [comment]
I just realized- StringEnums might break existing clients that expect the IntEnum value

File: Jellyfin.Server/Extensions/ApiApplicationBuilderExtensions.cs [hunk 1]
```diff
@@ -11,17 +12,38 @@ public static class ApiApplicationBuilderExtensions
...
             // specifying the Swagger JSON endpoint.
-            return applicationBuilder.UseSwaggerUI(c =>
+
+            var baseUrl = serverConfigurationManager.Configuration.BaseUrl.Trim('/');
+            if (!string.IsNullOrEmpty(baseUrl))
             {
-                c.SwaggerEndpoint("/swagger/v1/swagger.json", "Jellyfin API V1");
-            });
+                baseUrl += '/';
+            }
```
[the remark this answers is not in the document]
Comment (crobibero) [reply Jellyfin.Server/Extensions/ApiApplicationBuilderExtensions.cs:28]: RoutePrefix  is a relative path, and the json path is an absolute path, so I don't see a way to easily use the raw value.

[the remark this answers is not in the document]
Comment (mark-monteiro) [reply Jellyfin.Server/Extensions/ApiApplicationBuilderExtensions.cs:28]: A you're right, I missed the `RoutePrefix` bit. Looks good.

I might leave out the `baseUrl += '/';` bit here to make the URLs look a bit cleaner below, but that's not a big issue

[the remark this answers is not in the document]
Comment (crobibero) [reply Jellyfin.Server/Extensions/ApiApplicationBuilderExtensions.cs:28]: I would do that but then I would need more trickery for the RoutePrefix lines as they are relative :)

**Comment by crobibero** [comment]
> 
> 
> Tried this out locally and it's a great improvement! This will be awesome once we have everything working with the new ASP framework :)
> 
> One additional note, the swagger path is also used here and will need to be updated to match your change: https://github.com/jellyfin/jellyfin/blob/master/Jellyfin.Server/Program.cs#L530. Unfortunately I don't think you'll be able to access the base path at that point of the application startup. What I would recommend is removing that logic from `Program.cs` to and putting it in `HTTPListenerHost` where the `DefaultRedirectKey` setting is actually used and where you can also access the base URL

The `DefaultRedirectKey` automatically appends the path, so didn't need to move the logic to `HTTPListenerHost`

File: Jellyfin.Server/Extensions/ApiServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -105,6 +105,10 @@ public static IServiceCollection AddJellyfinApiSwagger(this IServiceCollection s
                 {
                     c.IncludeXmlComments(xmlFile);
                 }
+
+                // Order actions by route path, then by http method.
+                c.OrderActionsBy(description =>
+                    $"{description.ActionDescriptor.RouteValues["controller"]}_{description.HttpMethod}");
```
Comment (mark-monteiro) [inline Jellyfin.Server/Extensions/ApiServiceCollectionExtensions.cs:111]: Can you expand on what this changes? This seems sensible to me, but what was the original ordering?

Comment (crobibero) [reply Jellyfin.Server/Extensions/ApiServiceCollectionExtensions.cs:111]: The original ordering was whatever order the routes were created in the controller. Ideally they would also be created in the "correct" order, but this forces the swagger to generate it ordered.

I found this when I was reading up on swashbuckle: https://github.com/domaindrivendev/Swashbuckle.AspNetCore#change-operation-sort-order-eg-for-ui-sorting

Comment (mark-monteiro) [reply Jellyfin.Server/Extensions/ApiServiceCollectionExtensions.cs:111]: Cool, makes sense to me thanks! Ideally we would have a sensible ordering in our controllers but that seems like a real pain to decide on a standard for, code review, etc. So 👍 to this.

File: Jellyfin.Server/Extensions/ApiApplicationBuilderExtensions.cs [hunk 3]
```diff
@@ -12,17 +12,39 @@ public static class ApiApplicationBuilderExtensions
...
+            this IApplicationBuilder applicationBuilder,
+            IServerConfigurationManager serverConfigurationManager)
         {
-            applicationBuilder.UseSwagger();
-
             // Enable middleware to serve swagger-ui (HTML, JS, CSS, etc.),
             // specifying the Swagger JSON endpoint.
-            return applicationBuilder.UseSwaggerUI(c =>
+
+            var baseUrl = serverConfigurationManager.Configuration.BaseUrl.Trim('/');
```
Comment (Bond-009) [inline Jellyfin.Server/Extensions/ApiApplicationBuilderExtensions.cs:24]: Technically this should only be a `TrimStart`, but whatever

**Pull request merged**